File: tools/capture-weights/reference/owlv2_parity.ts
/**
 * OWLv2 base patch16 in the engine, finding four named things in a pinned real frame, held to its
 * upstream on the CPU and on the device at both precisions.
 *
 * **The engine half of the real model's parity.** `owlv2_parity.py` wrote the upstream's prepared
 * image, its token ids and its answers; this converts the same fetched checkpoint — verified by its
 * hash — into the image and text graphs, runs them over that very image, so preparation is not what
 * is compared, and reports how far each answer sits from the upstream's relative to its largest
 * value, how many detections agree, and how far their boxes land in the frame's own pixels.
 *
 * **The tokenizer is held to the upstream's exactly**, on the real vocabulary and on strings chosen
 * to be awkward: a token that differs is a query the text model was never asked.
 *
 * **It fails beyond a stated tolerance**, in `TOLERANCE`, each above what an RX 9070 XT measured —
 * the plan's Task 6 record has every figure.
 *
 *     env -u LD_LIBRARY_PATH .reference/venv-da3/bin/python tools/capture-weights/reference/owlv2_parity.py
 *     npx tsx --conditions=drift-source tools/capture-weights/reference/owlv2_parity.ts [--cpu] [--device]
 */
import { readFileSync } from 'node:fs';
import path from 'node:path';

import gpu from '@kmamal/gpu';

import {
  OWLV2_BASE,
  clipTokenizer,
  owlv2Detections,
  owlv2Image,
  owlv2Logits,
  owlv2Text,
  owlv2Tokens,
} from '../../../packages/capture/src/index.ts';
import { openInferenceDevice } from '../../../packages/core/src/render/inference/device.ts';
import {
  createGraphRunner,
  type GraphRunner,
} from '../../../packages/core/src/render/inference/runner.ts';
import {
  createGraphEvaluator,
  graphForDevice,
  type NetworkGraph,
} from '../../../packages/texture/src/index.ts';
import { tensorFloats } from '../checkpoint.ts';
import { clipMerges } from '../clipMerges.ts';
import { convert } from '../convert.ts';
import { readSafetensors } from '../safetensors.ts';

const ROOT = new URL('../../..', import.meta.url).pathname;
const only = process.argv.includes('--cpu')
  ? 'cpu'
  : process.argv.includes('--device')
    ? 'device'
    : 'all';
const folder = path.join(ROOT, '.reference', 'owlv2-parity');
const answers = readSafetensors(
  new Uint8Array(readFileSync(path.join(folder, 'output.safetensors'))),
);
const answer = (name: string): Float32Array => tensorFloats(name, answers.get(name) as never);
const asked = JSON.parse(readFileSync(path.join(folder, 'tokens.json'), 'utf8')) as {
  queries: string[];
  strings: Record<string, number[]>;
};
const [height, width] = Array.from(answer('original')) as [number, number];
const config = OWLV2_BASE;
const count = asked.queries.length;

const { models } = JSON.parse(
  readFileSync(path.join(ROOT, 'tools', 'capture-weights', 'manifest.json'), 'utf8'),
) as {
  models: {
    name: string;
    sha256: string;
    file: string;
    extra?: { file: string; sha256: string }[];
  }[];
};
const model = models.find((entry) => entry.name === 'owlv2-base-patch16');
if (model === undefined) throw new Error('the manifest has no owlv2-base-patch16');
const at = path.join(ROOT, 'models', 'capture', model.name);
const checkpoint = new Uint8Array(readFileSync(path.join(at, path.basename(model.file))));
const pinned = { sha256: model.sha256, format: 'safetensors' } as const;
const merges = clipMerges(
  readFileSync(path.join(at, 'merges.txt'), 'utf8'),
  JSON.parse(readFileSync(path.join(at, 'vocab.json'), 'utf8')) as Record<string, number>,
);
const beside = new Map([['tokenizer.merges', { shape: [merges.length / 2, 2], data: merges }]]);
const graphs = {
  image: convert(checkpoint, owlv2Image(config), pinned, beside),
  text: convert(checkpoint, owlv2Text(config, count), pinned, beside),
};

/**
 * Each answer relative to its largest value, a detection's box in the frame's pixels, and the
 * scores absolute — each about three times the worst an RX 9070 XT measured: **2.8e-4, 4.7e-3 px
 * and 1.2e-5 in single precision**, where the CPU's own reference measures 1.0e-4, 1.1e-3 px and
 * 4.1e-6 — so most of that is twelve layers of arithmetic rather than the device; and **9.8e-2,
 * 0.62 px and 3.1e-3 with the weights in half**, where every one of the 56 detections still
 * answers the same query. A tokenizer's ids are held exactly and have no tolerance.
 */
const TOLERANCE = {
  single: { value: 1e-3, box: 2e-2, score: 5e-5 },
  half: { value: 3e-1, box: 2, score: 1e-2 },
} as const;
type Limits = (typeof TOLERANCE)[keyof typeof TOLERANCE];
const failures: string[] = [];
const check = (label: string, value: number, limit: number): string => {
  if (!(value <= limit)) failures.push(`${label}: ${value.toExponential(2)} beyond ${limit}`);
  return value.toExponential(1);
};
const relative = (actual: Float32Array, expected: Float32Array): number => {
  let worst = 0;
  let scale = 0;
  for (let i = 0; i < expected.length; i += 1) {
    worst = Math.max(worst, Math.abs((actual[i] as number) - (expected[i] as number)));
    scale = Math.max(scale, Math.abs(expected[i] as number));
  }
  return worst / scale;
};

/* The tokenizer, on the real merges: every string's ids are the upstream's or the run fails. */
const tokenizer = clipTokenizer(merges);
for (const [text, ids] of Object.entries(asked.strings)) {
  const ours = tokenizer.encode(text);
  if (ours.length !== ids.length || ours.some((id, i) => id !== ids[i])) {
    failures.push(`tokenizer ${JSON.stringify(text)}: [${ours}] against [${ids}]`);
  }
}
const { tokens, ends } = owlv2Tokens(tokenizer, asked.queries, config.text.positions);
const upstreamTokens = answer('tokens');
for (let i = 0; i < upstreamTokens.length; i += 1) {
  if (tokens[i] !== upstreamTokens[i]) {
    failures.push(`query token ${i}: ${tokens[i]} against ${upstreamTokens[i]}`);
  }
}
console.log(
  `${Object.keys(asked.strings).length} strings and ${count} queries tokenized` +
    `${failures.length === 0 ? ' as the upstream does' : ''}`,
);

type Run = (
  graph: NetworkGraph,
  inputs: ReadonlyMap<string, Float32Array>,
) => Promise<Float32Array[]>;

async function compare(label: string, run: Run, limits: Limits): Promise<void> {
  const started = performance.now();
  const [classes, shift, scale, boxes, objectness] = (await run(
    graphs.image,
    new Map([['image', answer('image')]]),
  )) as [Float32Array, Float32Array, Float32Array, Float32Array, Float32Array];
  const [queries] = (await run(
    graphs.text,
    new Map([
      ['tokens', tokens],
      ['ends', ends],
    ]),
  )) as [Float32Array];
  const logits = new Float32Array(shift.length * count);
  owlv2Logits({ classes, shift, scale }, queries, count, logits);

  const lines = [];
  for (const [name, values] of [
    ['classes', classes],
    ['boxes', boxes],
    ['objectness', objectness],
    ['queries', queries],
    ['logits', logits],
  ] as const) {
    const off = relative(values, answer(name));
    lines.push(`${name} ${check(`${label} ${name}`, off, limits.value)}`);
  }
  const found = owlv2Detections(logits, boxes, count, 0.1, height, width);
  const scores = answer('detections.scores');
  const labels = answer('detections.labels');
  const upstreamBoxes = answer('detections.boxes');
  let score = 0;
  let box = 0;
  let wrong = found.length === scores.length ? 0 : 1;
  found.forEach((detection, i) => {
    if (detection.label !== labels[i]) wrong += 1;
    score = Math.max(score, Math.abs(detection.score - (scores[i] as number)));
    detection.box.forEach((value, j) => {
      box = Math.max(box, Math.abs(value - (upstreamBoxes[i * 4 + j] as number)));
    });
  });
  if (wrong > 0)
    failures.push(
      `${label}: ${found.length} detections against ${scores.length}, ${wrong} differing`,
    );
  console.log(
    `${label} (${((performance.now() - started) / 1000).toFixed(1)} s)\n  ${lines.join('; ')}\n` +
      `  ${found.length} detections, scores ${check(`${label} scores`, score, limits.score)}, ` +
      `boxes ${check(`${label} boxes`, box, limits.box)} px`,
  );
}

console.log(
  `a ${width}×${height} frame at ${config.size}², ${count} queries, against the upstream in single precision`,
);
if (only !== 'device') {
  await compare(
    'cpu',
    async (graph, inputs) => {
      const out = createGraphEvaluator(graph).run(inputs);
      return graph.outputs.map((name) => Float32Array.from(out.get(name) as Float32Array));
    },
    TOLERANCE.single,
  );
}
if (only !== 'cpu') {
  const instance = gpu.create([]) as unknown as GPU;
  for (const half of [false, true]) {
    const inference = await openInferenceDevice(instance, { half });
    const runners = new Map<NetworkGraph, GraphRunner>();
    await compare(
      `device, ${half ? 'half' : 'single'}`,
      async (graph, inputs) => {
        let runner = runners.get(graph);
        if (runner === undefined) {
          runner = await createGraphRunner(inference, graphForDevice(graph));
          runners.set(graph, runner);
        }
        const out = await runner.run(inputs);
        return graph.outputs.map((name) => Float32Array.from(out.get(name) as Float32Array));
      },
      half ? TOLERANCE.half : TOLERANCE.single,
    );
    for (const runner of runners.values()) runner.dispose();
  }
}
if (failures.length > 0) console.log(`\noutside the tolerance:\n  ${failures.join('\n  ')}`);
/* Dawn asserts on a natural exit over objects already released; see the native host's `game.ts`. */
process.exit(failures.length === 0 ? 0 : 1);

File: tools/capture-weights/reference/owlv2_parity.py
"""OWLv2 base patch16, as its upstream runs it, finding four named things in a pinned real frame.

The upstream half of the real model's parity: this loads the fetched checkpoint into Transformers'
`Owlv2ForObjectDetection` at the manifest's pinned revision, refusing anything missing or left
over, prepares the first of Depth Anything 3's two pinned frames and four text queries with the
upstream's own processor -- the image padded to a square after its end, blurred against aliasing and
resized to 960, normalised by CLIP's statistics; each query tokenized and padded to 16 -- and writes
the prepared image, the token ids, what the model computes before its class head joins them, its
answers and its detections to `.reference/owlv2-parity/`, where `owlv2_parity.ts` holds the engine to
them, on the CPU and on the device. **And the tokenizer's answer for a list of awkward strings**, so
the engine's tokenizer is held to the upstream's on the real vocabulary.

    node tools/capture-weights/fetch.mjs owlv2-base-patch16
    env -u LD_LIBRARY_PATH .reference/venv-da3/bin/python tools/capture-weights/reference/owlv2_parity.py

The model folder is the fetched checkpoint beside the configuration, processor and tokenizer files
at the same revision, under `.reference/owlv2/model/`.
"""

import hashlib
import json
from pathlib import Path

import numpy as np
import torch
from PIL import Image
from safetensors.torch import save_file
from transformers import Owlv2ForObjectDetection, Owlv2Processor

ROOT = Path(__file__).resolve().parents[3]
FOLDER = ROOT / ".reference" / "owlv2-parity"
FOLDER.mkdir(exist_ok=True)
MODEL = ROOT / ".reference" / "owlv2" / "model"
manifest = json.loads((ROOT / "tools" / "capture-weights" / "manifest.json").read_text())
entry = next(m for m in manifest["models"] if m["name"] == "owlv2-base-patch16")
if hashlib.sha256((MODEL / "model.safetensors").read_bytes()).hexdigest() != entry["sha256"]:
    raise SystemExit("the checkpoint is not the one the manifest pins; fetch it again")
for extra in entry["extra"]:
    if hashlib.sha256((MODEL / extra["file"]).read_bytes()).hexdigest() != extra["sha256"]:
        raise SystemExit(f"{extra['file']} is not the one the manifest pins")
image = json.loads((Path(__file__).parent / "inputs.json").read_text())["images"][0]
path = ROOT / ".reference" / "da3-parity" / image["file"]
if hashlib.sha256(path.read_bytes()).hexdigest() != image["sha256"]:
    raise SystemExit(f"{path} is not the frame the reference pins")
frame = Image.open(path).convert("RGB")

model, info = Owlv2ForObjectDetection.from_pretrained(MODEL, output_loading_info=True)
problems = {key: value for key, value in info.items() if value}
if problems:
    raise SystemExit(f"the checkpoint does not load as the upstream's model: {problems}")
model.eval()
processor = Owlv2Processor.from_pretrained(MODEL)

QUERIES = ["a skyscraper", "a bridge", "a boat", "a tree"]
# Strings for the tokenizer alone: case, runs of space, contractions, digits, punctuation runs,
# accents composed and not, a soft hyphen, the padding token inside a word, the specials by name,
# and a word the merges build in several steps.
STRINGS = [
    "a photo of a cat", "  A   Photo\tOF a\nCAT  ", "it's the dog's bone, isn't it?", "route 66 and 2024",
    "...!!!???", "café, café and CAFÉ", "soft­hyphen", "wow! great!", "<|startoftext|>a<|endoftext|>b",
    "<|StartOfText|>c", "antidisestablishmentarianism", "ΣΟΦΊΑ σοφία", "東京タワー", "emoji 🚲 bike",
]

with torch.inference_mode():
    tokens = {s: processor.tokenizer(s)["input_ids"] for s in STRINGS}
    inputs = processor(text=[QUERIES], images=frame, return_tensors="pt")
    captured = {}

    def keep(name):
        def hook(_module, _inputs, output):
            captured[name] = output.detach().clone()

        return hook

    model.class_head.dense0.register_forward_hook(keep("classes"))
    model.class_head.logit_shift.register_forward_hook(keep("shift"))
    model.class_head.logit_scale.register_forward_hook(keep("scale"))
    model.owlv2.text_projection.register_forward_hook(keep("queries"))
    out = model(**inputs)
    found = processor.post_process_grounded_object_detection(
        out, threshold=0.1, target_sizes=[(frame.height, frame.width)], text_labels=[QUERIES]
    )[0]

rgb = np.array(frame)
answers = {
    "image": inputs["pixel_values"][0].contiguous(),
    # The frame as a host hands it to the engine, for the engine's own preparation to be held to the
    # processor's above.
    "frame.rgba": torch.tensor(
        np.concatenate([rgb, np.full(rgb.shape[:2] + (1,), 255, dtype=np.uint8)], axis=-1).astype(np.float32)
    ),
    "tokens": inputs["input_ids"].to(torch.float32).contiguous(),
    "original": torch.tensor([frame.height, frame.width], dtype=torch.float32),
    "classes": captured["classes"][0].contiguous(),
    "shift": captured["shift"][0].contiguous(),
    "scale": captured["scale"][0].contiguous(),
    "boxes": out.pred_boxes[0].contiguous(),
    "objectness": out.objectness_logits[0].contiguous(),
    "queries": captured["queries"].contiguous(),
    "logits": out.logits[0].contiguous(),
    "detections.scores": found["scores"].contiguous(),
    "detections.labels": found["labels"].to(torch.float32).contiguous(),
    "detections.boxes": found["boxes"].contiguous(),
}
save_file(answers, FOLDER / "output.safetensors")
(FOLDER / "tokens.json").write_text(json.dumps({"queries": QUERIES, "strings": tokens}, indent=2, ensure_ascii=False))
print(f"{len(found['scores'])} detections above 0.1 on the {frame.width}×{frame.height} frame")
for score, label, box in zip(found["scores"], found["text_labels"], found["boxes"]):
    print(f"  {label}: {score.item():.3f} at {[round(b.item(), 1) for b in box]}")
